### PR TITLE
feat: volume discount program

### LINKED ISSUE #9302: [Feature]: Volume Discount Program
Implement volume discount as specified in https://github.com/vegaprotocol/specs/blob/cosmicelevator/protocol/0084-VDPR-volume_discount_program.md

### PR DESCRIPTION
**Pull request by ze97286**
Closes #9302

### CODE REVIEW

File: datanode/gateway/graphql/schema.graphql [hunk 1]
```diff
@@ -4096,6 +4096,60 @@ type UpdateMarketState {
   price: String
 }
 
+
+type UpdateReferralProgram {
+  "The current version of the referral program"
+  version: Int!
+  "The identifier of the proposal that created the referral program"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:4103]:
```suggestion
  "ID of the proposal that created the referral program"
```

File: datanode/gateway/graphql/schema.graphql [hunk 2]
```diff
@@ -4096,6 +4096,60 @@ type UpdateMarketState {
   price: String
 }
 
+
+type UpdateReferralProgram {
+  "The current version of the referral program"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:4101]:
```suggestion
  "Current version of the referral program"
```

File: datanode/gateway/graphql/schema.graphql [hunk 3]
```diff
@@ -4096,6 +4096,60 @@ type UpdateMarketState {
...
+  version: Int!
+  "The identifier of the proposal that created the referral program"
+  id: ID! 
+  "The benefit tears for the program"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:4105]:
```suggestion
  "Benefit tiers for the program"
```

File: datanode/gateway/graphql/schema.graphql [hunk 4]
```diff
@@ -4096,6 +4096,60 @@ type UpdateMarketState {
...
+  "The window legnth to consider for the referral program"
+  windowLength: Int!
+}
+
+type BenefitTier {
+  "The minimum number of epochs the party needs to be in the referral set to be eligible for the benefit"
+  minimumEpochs: Int!
+  "The minimum running notional for the given benefit layer"
+  minimumRunningNotionalTakerVolume: String! 
+  "The roportion of the referee's taker fees to be discounted"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:4120]:
```suggestion
  "The proportion of the referee's taker fees to be discounted"
```

File: datanode/gateway/graphql/schema.graphql [hunk 5]
```diff
@@ -4096,6 +4096,60 @@ type UpdateMarketState {
...
+  referralDiscountFactor: String!
+  "The roportion of the referee's taker fees to be rewarded to the referrer"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:4122]:
```suggestion
  "The proportion of the referee's taker fees to be rewarded to the referrer"
```

File: datanode/gateway/graphql/schema.graphql [hunk 6]
```diff
@@ -4096,6 +4096,60 @@ type UpdateMarketState {
...
+  "Required number of governance tokens ($VEGA) a referrer must have staked to receive the multiplier"
+  minimumStakedTokens: String!
+  "Multiplier applied to the referral reward factor when calculating referral rewards due to the referrer"
+  referralRewardMultiplier: String!
+}
+
+type UpdateVolumeDiscountProgram {
+  "The current version of the volume discount program"
+  version: Int!
+  "The identifier of the proposal that created the discount program"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:4136]:
```suggestion
  "ID of the proposal that created the discount program"
```

File: datanode/gateway/graphql/schema.graphql [hunk 7]
```diff
@@ -4096,6 +4096,60 @@ type UpdateMarketState {
...
+  "The end time of the program"
+  endOfProgramTimestamp: Timestamp!
+  "The window legnth to consider for the volume discount program"
+  windowLength: Int!
+}
+
+type VolumeBenefitTier {
+  "The minimum running notional for the given benefit tier"
+  minimumRunningNotionalTakerVolume: String! 
+  "The discount given for this layer"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:4149]:
```suggestion
  "Discount given to those in this benefit tier"
```

Comment (candida-d) [reply datanode/gateway/graphql/schema.graphql:4149]: Does this change accurately describe it?

File: protos/sources/vega/governance.proto [hunk 8]
```diff
@@ -553,6 +557,10 @@ message Vote {
   string total_equity_like_share_weight = 7;
 }
 
+message UpdateVolumeDiscountProgram {
+  VolumeDiscountProgram changes = 1;
```
Comment (candida-d) [inline protos/sources/vega/governance.proto:561]:
```suggestion
 // Configuration for a change to update a volume discount program
  VolumeDiscountProgram changes = 1;
```

File: protos/sources/vega/vega.proto [hunk 9]
```diff
@@ -1739,6 +1747,23 @@ message StakingTier {
...
+  // update.
+  uint64 version = 1;
+  // Unique ID generated from the proposal that created this program.
+  string id = 2;
+  // Defined benefit tiers in increasing order. First element will give Tier 1,
+  // second element will give Tier 2, and so on. Determines the level of
+  // benefit a party can expect based on performance criteria.
+  repeated VolumeBenefitTier benefit_tiers = 3;
+  // Timestamp as Unix time in seconds after which when the current epoch ends, the
+  // programs status will become STATE_CLOSED and benefits will be disabled.
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:1761]:
```suggestion
  // program's status will become STATE_CLOSED and benefits will be disabled.
```

**Review by candida-d — APPROVED** [review]

**Review by jeremyletang — APPROVED** [review]

**Review by cdummett — APPROVED** [review]

**Review by davidsiska-vega — APPROVED** [review]
Looked at `defaults.go` and the limits / default values look good... though we may want to update the default values later to what we want the upgrade of mainnet to initialise with.

**Pull request merged**